safe_control_gym/experiments/epoch_experiment.py

```diff
@@ -0,0 +1,226 @@
+                                  **kwargs):
+        """Defined per controller?"""
+        # Training Data Collection
+        traj_data = self._execute_task(ctrl=run_ctrl,
+                                       env=env,
+                                       n_episodes=n_episodes,
+                                       n_steps=n_steps,
+                                       log_freq=log_freq)
+        # Add trajectory data to all training data.
+        if self.save_train_trajs:
+            self.add_to_all_train_data(traj_data)
+        # Parsing of training Data.
+        train_inputs, train_outputs = self.preprocess_training_data(traj_data, **kwargs)
+        # Learning of training data.
+        self.train_controller(train_inputs, train_outputs, **kwargs)
+
+    def preprocess_training_data(self, traj_data, **kwargs):
+        """Preprocess the trajectory data into training data for controller training.
+
+        Args:
+            traj_data (dict): Dictionary of the trajectory data.
+
+        """
+
+        raise NotImplementedError('preprocessing_training_data not implemented!')
+
+    def train_controller(self, train_inputs, train_outputs, **kwargs):
+        """Train the controller from training data train_inputs and train_outputs. The intention is that
+        preprocess_trianing_data has processed the data and formatted such that ctrl.learn should be able to learn
+        from the given data.
+
+        Args:
+            train_inputs (np.array): Training data array.
+            train_outputs (np.array): Training data targets.
+        """
+        train_output = self.ctrl.learn(input_data=train_inputs, target_data=train_outputs)
+        raise NotImplementedError('Controller training needs to be implemented.')
+
+
+    def add_to_all_train_data(self, traj_data):
+        if self.all_train_data is None:
+            self.all_train_data = deepcopy(traj_data)
+        else:
+            for key in self.all_train_data:
+                self.all_train_data[key].append(deepcopy(traj_data[key][0]))
+
+    def add_to_all_test_data(self, traj_data):
+        if self.all_test_data is None:
+            self.all_test_data = deepcopy(traj_data)
+        else:
+            for key in self.all_test_data:
+                self.all_test_data[key].append(deepcopy(traj_data[key][0]))
+
+    def add_to_metrics(self, metrics):
+        if self.metrics is None:
+            self.metrics = deepcopy(metrics)
+        else:
+            for key in self.metrics:
+                self.metrics[key].append(deepcopy(metrics[key][0]))
+    def compute_metrics(self, trajs_data):
+        '''Compute all standard metrics on the given trajectory data.
+
+        Args:
+            trajs_data (defaultdict(list)): The raw data from the executed runs.
+
+        Returns:
+            metrics (dict): The metrics calculated from the raw data.
+        '''
+
+        metrics = super().compute_metrics(trajs_data)
+        list_metrics = {}
```

Comment:
Does this just copy the dict? What does this function do differently than the existing compute metrics?

Reply:
The `BaseExperiment` class is simply a dictionary with `key: <value>`. Here, we slightly modified it such that it is now 'key: [<value1>, <value2>,...]` so we can compute the metrics for each epoch.